ktor-utils/jvm/src/io/ktor/util/reflect/ServiceLoader.kt

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.util.reflect
+
+import io.ktor.utils.io.*
+import java.util.*
+
+/**
+ * Loads implementations of service [T] using [ServiceLoader.load].
+ *
+ * NOTE: ServiceLoader should use specific call convention to be optimized by R8 on Android:
+ * `ServiceLoader.load(X.class, X.class.getClassLoader()).iterator()`
+ * See: https://r8.googlesource.com/r8/+/refs/heads/main/src/main/java/com/android/tools/r8/ir/optimize/ServiceLoaderRewriter.java
+ */
+@InternalAPI
+public inline fun <reified T> loadService(): Iterable<T> = Iterable {
```

Comment:
note: it's better to recheck, that R8 will optimize this pattern with `inline` and `reified` usage. While from the code perspective it should be the same, I'm not confident, that the byte code will be the same.

Reply:
Yes, I'm not sure either 😅 I'm going to test it

Reply:
Yes, the optimization works!

During the testing I've found a more serious startup performance issue not related to `ServiceLoader`. We call `typeOf` during initialization of each plugin, and it is extremely slow when user have `kotlin-reflect` in dependencies.
Unoptimized service loader call takes ~20ms while initialization of HttpClient with five plugins installed takes more than a second! 😱 
I'll create a new issue for this problem.

Reply:
New issue:
- [KTOR-7715](https://youtrack.jetbrains.com/issue/KTOR-7715) Plugins’ initialization is slow on Android if kotlin-reflect is present